### ISSUE TITLE
Metering of stream's audiotrack?

### ISSUE THREAD

**Issue by chadobado** [react-native-webrtc/react-native-webrtc#10]
Hey @oney, burned through a few hours on this - maybe it's something simple I'm missing.  Thanks in advance for any help.

I'm trying to do simple metering/measuring of audio from connected peers to visualize audio levels.  Pretty common usecase for video-chat applications.

Here are 3 approaches I've tried w/ no success:

**Polling/subscribing to microphone activity of the WebRTC MediaStream audiotrack**

Attempting to use audioContext (which appears the [correct approach](http://patdunn.net/2015/05/15/detect-and-visualize-whos-talking-in-webrtc/)), the web audio api (not supported in RN), [peerConnection.getStats()](http://www.callstats.io/2015/07/06/basics-webrtc-getstats-api/), RTCStatsReport, etc.

**Measuring audio played through the device speaker**

Explored trying to meter the AVAudioSession implementation in WebRTCModule+RTCSetting.m with AVCaptureSession (although I don't want to record anything, just measure it), using the AVAudioSessionModeMeasurement mode.

Also found some examples using AVAudioPlayer and AVAudioRecorder's built in metering methods, but all examples are geared towards playing a file vs a stream (via AVAudioSession).

**Creating a native module**

Bridge to an existing library w/ simplified audio handling and replace AVAudioSession references in WebRTCModule+RTCSetting.m.  Seems like overkill just to get audio levels.

Such as:  
- [EZAudio](https://github.com/syedhali/EZAudio)
- [StreamingKit](https://github.com/tumtumtum/StreamingKit)
- [Amazing audio engine](http://theamazingaudioengine.com/)

Any idea which approach is best here, or possibly a better alternative?

Thanks again for the help.

**Comment by chadobado** [comment]
Made some progress exposing a native module.  Am sure there are some mistakes.  

Not having written a native module before, it looks like the next step should be to pass a callback along with startMetering that is fired by NSTimer/updateAudioMeter, pushing the volume level back to JS so react can setState({volume:value}) and update views accordingly.

If you get a chance, love to know if I'm barking up the wrong tree.  Thanks.

```objc
RCT_EXPORT_METHOD(startMetering:(BOOL)isOn)
{
  NSLog(@"Starting audio metering");
  [[AVAudioSession sharedInstance] setCategory: AVAudioSessionCategoryPlayAndRecord error: nil];

  NSURL *url = [NSURL fileURLWithPath:@"/dev/null"];  //Don't actually want to record anything
  
  NSDictionary *recordSettings =
  [[NSDictionary alloc] initWithObjectsAndKeys:
  [NSNumber numberWithFloat: 44100.0], AVSampleRateKey,
  [NSNumber numberWithInt: kAudioFormatAppleLossless], AVFormatIDKey, [NSNumber numberWithInt: 1], AVNumberOfChannelsKey, [NSNumber numberWithInt: AVAudioQualityMax], AVEncoderAudioQualityKey, nil];

  AVAudioRecorder *audioRecorder = [[AVAudioRecorder alloc] initWithURL: url settings:recordSettings error:nil];

  [audioRecorder prepareToRecord];
  audioRecorder.meteringEnabled = TRUE;
  [audioRecorder record];
  
  NSTimer *meterTimer = [NSTimer scheduledTimerWithTimeInterval:0.5 target:self selector:@selector(updateAudioMeter:) userInfo:audioRecorder repeats: YES];

}

- (void)updateAudioMeter:(NSTimer *)timer
{
  AVAudioRecorder *audioRecorder = [timer userInfo];
  [audioRecorder updateMeters];

  //exploring audio channels:
  for (int k=0; k < 2; k++) {
    float peak = [audioRecorder peakPowerForChannel:k];
    float average = [audioRecorder averagePowerForChannel:k];
    NSLog(@"Peak power for channel %i: %4.2f",k,peak);
    NSLog(@"Average power for channel %i:%4.2f",k,average);
  }
  
  //realized all my audio is on channel0:
  float level = [audioRecorder peakPowerForChannel:0];
  NSLog(@"Db: %f", level);

  double percentage = pow (10, (0.05 * level));
  
  NSLog(@"Percent: %f", percentage);
  
  
 
}
```

**Comment by oney** [comment]
I think exposing EZAudio or `audioRecorder updateMeters` to JS is right way.
Besides, we have [`RTCStatsDelegate`](https://github.com/oney/react-native-webrtc/blob/master/RCTWebRTC/libjingle_peerconnection/Headers/RTCStatsDelegate.h#L34) on Objective-C corresponding with `peerConnection.getStats()`, so bridging it to JS is also an approach.

**Comment by chadobado** [comment]
Great thanks for the feedback, I'll circle back when I have something that can benefit others.

Happy holidays!

**Comment by oney** [comment]
You, too!
I will need this feature for my app, but I am currently busy at work. Looking forward to your PR, it will be very useful to me.

**Comment by jsdario** [comment]
Did this move any forward?

**Comment by Ulydev** [comment]
Hi, has there been any progress with this? I'm wondering if after more than two years this has been solved or not, as this is a pretty common feature for video/voice chat applications :-)

**Comment by jadbox** [comment]
I really need this feature too :/

**Comment by jadbox** [comment]
@chadobado any progress on this front?

**Comment by saghul** [comment]
@jadbox I believe the is: getStats. However, as I mentioned in #276 it does have a performance impact if called too often.

**Comment by ajnozari** [comment]
So going over the documentation you could do a simple math equation every couple of seconds

Math.sqrt(totalAudioEnergy/totalSamplesDuration), this ends up giving the same units as audioLevel. The duration can be the entirety of the call, OR just a short 3-5 seconds to ms, if you want to get a change over time (detect who's talking) or if you want a specific last few seconds range Math.sqrt((energy2 - energy1)/(duration2 - duration1)).

This is the same equation that is done to get audioLevel in later implementations of webRTC, which is a 0-1 scale (0 = silence, 1 = the loudest the device can theoretically pick up.  The closer to 1 you are the louder, 0.5 is therefore ~ 6 dBSPL). 

You'll need to play with it, but if you set your duration small enough (I use 1 second, you can go smaller) you'll find that you get a reasonable approximation that can be used to detect who's the loudest in a group of people. Couple that with a timeout, lock variable, and a "who's currently speaking id" and you can give yourself a "who's talking" and if you multiply by 100 you *could* implement this scale as a percent and use it for an audio visualizer.

Just remember as @saghul said, don't run the interval too often or you'll wreck performance. Try to run it no more than every 2 seconds, which I found was a happy medium even when running in dev mode. I wouldn't do less often if you want a visualizer, but since my focus is "who's talking" I don't need to run it as often and could probably get away with 3-5 seconds since I lockout anyways and only clear the lockout if they fall below .1.

**Comment by alenthomas343** [comment]
Hello sir,
I am new to react native and I wish to develop a small video call app with react native. I have made some small apps, but I need to integrate webrtc for video call in my app. Can you please help me because I am facing some issue relation with integrating it my video is not at all displaying in other device ie, i am working with android devices. Can you please help me with my issue.

**Comment by ajnozari** [comment]
We ended up implementing JanusGateway, although first we did get the webrtc working.

Are you sure your signaling to the device you’re trying to connect to is working?  Are you also sure the correct device is responding?  If you have multiple devices that can cause an issue where the wrong device tries to respond.

**Comment by Dajust** [comment]
@ajnozari Just curious, why did you go with JanusGatway although you got webrtc working?

**Comment by ajnozari** [comment]
So what Janus provided for us was rather unique.  

First off it reduced the power from handling multiple video participants, and allowed for better bandwidth control which helped users in constrained connections/older devices. 

It also assisted in nat traversal since it is a single, non-nat’d source for all connections (send and receive), eliminating most firewall issues we’ve had.

Even better it enabled active speaker, which made small screen speaker swapping easier since RN-webrtc doesn’t expose the newer audio level api. Before Janus we were relying on comparing relative audio volume from each stream to decide who was talking, this didn’t always work.  That being said Janus’ implementation still relies on devices reporting this, and while it works on iOS and macOS, android and web surprisingly is stubborn.

Further, it allowed us to offload recording from the local device where incoming streams are affected by bandwidth in, to a single source server that provided much more consistent and reliable recordings.

Lastly, it made screen sharing/stream swapping issues a thing of the past.  It uses a separate stream for screenshare.  RN has an issue where it doesn’t switch the video streams reliably as it doesn’t listen to the signal when switching from a desktop/app screenshare back to your webcam. This caused black screens and required renegotiating the video stream.... every switch.

Janus has served us well and has essentially eliminated every issue above while adding features we don’t quite use yet, but could easily implement in the future with ease.

Tldr: it was an interesting interplay of “what can be run cost effectively in the cloud that will improve the end users experience”.

Note:  I don’t work for Janus, their product just solved every issue we had and was open source.

**Comment by Dajust** [comment]
Wow, thanks for the detailed response. I'm still working on my web app and I'm just about to use [hark](https://github.com/otalk/hark) for voice detection, which apparently doesn't work for RN. Now I'm not sure how to continue this RTC feature because I really needed to reuse the same code for RN. Or, I might just rewrite the thing using Janus when I'm ready for the native app.

**Comment by minhchienwikipedia** [comment]
Any update?

**Comment by aprilmintacpineda** [comment]
also wanted to do this.
